### ISSUE TITLE
Function with missing case is reported as total

### ISSUE THREAD

**Issue by tsani** [idris-lang/Idris-dev#3942]
Consider this module:

```idris
module Bug

%default total

data Recognizer : tok -> Bool -> Type where
  Empty : Recognizer tok True
  Seq : Recognizer tok l ->
        Recognizer tok r ->
        Recognizer tok (l && r)

namespace Recognition
  data Recognition : List tok -> Recognizer tok n -> Type where
    Empty : {tok : Type} -> Recognition (the (List tok) Nil) Empty

    Seq : Recognition ts1 left -> Recognition ts2 right ->
          Recognition (ts1 ++ ts2) (Seq left right)

emptyMeansNull : {p : Recognizer tok n} -> Recognition [] p -> Nat
emptyMeansNull Empty = 5
```

The function `emptyMeansNull` is missing a case for the `Seq` constructor, but is reported as total. I can't strip this example down any more and reproduce the bug: removing the `List tok` parameter from `Recognition` or removing the `tok` parameter from `Recognizer` makes Idris correctly report the function as not total.

I'm using Idris 1.0.

**Comment by ahmadsalim** [comment]
Thanks for reporting the issue.
I think I have seen a similar issue before, but I can not remember where. If anyone else remembers please do remind me.

As for a workaround, you should use a variable and an equality constraint to delay the unification, e.g:

```idris
emptyMeansNull : {xs : List tok} -> {p : Recognizer tok n} -> (xs = []) -> Recognition xs p -> Nat
emptyMeansNull Refl Empty = 5
emptyMeansNull isEmpty (Seq r1 r2) = ?p 
-- use isEmpty here to discharge other cases than the one which has index []
```

If you want you can use this as a helper function instead and still have the same type signature for `emptyMeansNull` which then delegates to this helper.

**Comment by yanok** [comment]
Related to #4282 and #4299. Fixed by #4303.
